src/session.rs

```diff
@@ -26,65 +28,68 @@ use zenoh_util::core::zresult::ErrNo;
 /// After a move, `val` will still exist, but will no longer be valid. The destructors are double-drop-safe, but other functions will still trust that your `val` is valid.  
 ///
 /// To check if `val` is still valid, you may use `z_X_check(&val)` or `z_check(val)` if your compiler supports `_Generic`, which will return `true` if `val` is valid.
-#[cfg(not(target_arch = "arm"))]
-#[repr(C, align(8))]
-pub struct z_owned_session_t([u64; 3]);
-
-#[cfg(target_arch = "arm")]
-#[repr(C, align(4))]
-pub struct z_owned_session_t([u32; 3]);
+#[repr(C)]
+pub struct z_owned_session_t(usize);
 
-impl_guarded_transmute!(Option<Session>, z_owned_session_t);
+impl_guarded_transmute!(Option<Arc<Session>>, z_owned_session_t);
 
-impl From<Option<Session>> for z_owned_session_t {
-    fn from(val: Option<Session>) -> Self {
+impl From<Option<Arc<Session>>> for z_owned_session_t {
+    fn from(val: Option<Arc<Session>>) -> Self {
         val.transmute()
     }
 }
 
-impl AsRef<Option<Session>> for z_owned_session_t {
-    fn as_ref(&self) -> &Option<Session> {
+impl AsRef<Option<Arc<Session>>> for z_owned_session_t {
+    fn as_ref(&self) -> &Option<Arc<Session>> {
+        unsafe { std::mem::transmute(self) }
+    }
+}
+
+impl AsMut<Option<Arc<Session>>> for z_owned_session_t {
+    fn as_mut(&mut self) -> &mut Option<Arc<Session>> {
         unsafe { std::mem::transmute(self) }
     }
 }
 
-impl AsMut<Option<Session>> for z_owned_session_t {
-    fn as_mut(&mut self) -> &mut Option<Session> {
+impl AsRef<Option<Weak<Session>>> for z_session_t {
```

Comment:
I wonder if `Option<Weak<T>>` has any advantage vs `Weak<T>`. To use `Weak<T>`, `upgrade()` needs to be called in any case.

Reply:
That's to deal with the fact that `z_owned_session_t == Option<Arc<Session>>`, can't construct a `Weak` from a `None`.

But you've just made me realize something bad: since `z_session_t` is a `Option<Weak<Session>>`, it will keep the weak count up, which will prevent deallocation (the session _will_ be destructed when the strong count reaches 0). To deal with this, we have 3 choices:
- give `z_session_t` a destructor, which is senseful, but might be very confusing.
- When constructing `z_session_t` in `z_loan`, decrement the weak count manually. This does mean that `z_session_t` will be a `Weak` to deallocated memory, causing segfaults if used after all arcs to the session are destroyed.
- remove `z_session_t` entirely, since everything that uses the session does so immutably anyway. But it is an API-break

Reply:
I'm not sure to fully understand your first point. A `Weak` can be constructed without requiring passing from an `Arc`. It's enough to do a `Weak::new()` (see https://doc.rust-lang.org/std/sync/struct.Weak.html#method.new).

Regarding your second point, the strong count should not include the weak count (see https://doc.rust-lang.org/std/sync/struct.Arc.html#method.strong_count). Is there anything am I missing here?

Reply:
After internal discussion: the selected implementation is to have `z_session_t` be a raw pointer to the refcounted allocation, which means it is only valid as long as the refcount doesn't reach 0.